Added milestone counting support for maximum likelihood derived estimators

marscher opened the pull request:
* new msm API method "estimate_core_markov_model", dtrajs and core_set definition

TODO:
- [x] add tests
- [x] add references and examples to API docstring
- [ ] add core_set param to implied_timescales

jchodera commented:
Tagging @maxentile who may be interested

franknoe commented:
Thanks, I'm happy to review this. I suggest the following release plan:

* 2.3: Koopman reweighting stuff (almost done)
* 2.4: Core MSMs and maybe other things.

I.e. don't merge the core MSM stuff before the Koopman stuff is released.

marscher commented:
It is not yet done. Eg. tests are missing. I've discussed with Fabian, that we could use Koopman as a cross-validation, but we need to fiddle out the details.

For now I'm planning a test case in which we compare a standard msm on a kmeans discretization with a CoreMSM living on a density based disc and show that the
timescales are approximately equal for a high lag time of the standard and a low lag time for the core model.

franknoe commented:
ok, thanks.

Am 14/12/16 um 17:57 schrieb Martin K. Scherer:
> It is not yet done. Eg. tests are missing. I've discussed with Fabian, 
> that we could use Koopman as a cross-validation, but we need to fiddle 
> out the details.
>
> For now I'm planning a test case in which we compare a standard msm on 
> a kmeans discretization with a CoreMSM living on a density based disc 
> and show that the
> timescales are approximately equal for a high lag time of the standard 
> and a low lag time for the core model.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1000#issuecomment-267089624>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMeQqZLaHjtoJHHnzyXNDvOe1wLDiFfks5rIB_7gaJpZM4LAcej>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

franknoe commented:
OK, then I wait and it would be good if @fabian-paul  has a look at it first once it's done.
It would be good to have an example that somehow involves density-based clustering, to see how this fits together.

franknoe commented:
Let's plan the work on this core set stuff with @giopina next week.

marscher commented:
@giopina suggested a neat solution for the problem of a trajectory not starting in a core, which then has to be treated. We basically assign these staring pieces with an artificial state, which will then be disconnected and omitted by our MSM estimators. This avoids the problem of remapping the lengths of the trajectory, when we extract representative structures from the original MD trajectory.
As a convention for this artificial state we can use N_cores + 1. 

Currently I return an index offset map to solve this, but I find the above solution is better suited, because it is simple.

I suggest to have an API function in the msm package to perform milestone counting on the dtrajs list, which can then be passed (in principle) to any estimator.

franknoe commented:
An alternative is to do the milestoning assignment during the estimation 
process, i.e. in a CoreMSMEstimator. Then we can leave all non-core 
timeframes with -1 and keep the time assignment. The CoreMSMEstimator 
needs to change a few things e.g. in the frame sampling to exclude the 
-1 states.

Am 24/01/17 um 13:42 schrieb Martin K. Scherer:
>
> @giopina <https://github.com/giopina> suggested a neat solution for 
> the problem of a trajectory not starting in a core, which then has to 
> be treated. We basically assign these staring pieces with an 
> artificial state, which will then be disconnected and omitted by our 
> MSM estimators. This avoids the problem of remapping the lengths of 
> the trajectory, when we extract representative structures from the 
> original MD trajectory.
> As a convention for this artificial state we can use N_cores + 1.
>
> Currently I return an index offset map to solve this, but I find the 
> above solution is better suited, because it is simple.
>
> I suggest to have an API function in the msm package to perform 
> milestone counting on the dtrajs list, which can then be passed (in 
> principle) to any estimator.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1000#issuecomment-274793286>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMeQov3VR8KXZIT-_gfcB5bpJh2jx10ks5rVfG4gaJpZM4LAcej>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

marscher commented:
On 01/24/2017 01:46 PM, Frank Noe wrote:
> An alternative is to do the milestoning assignment during the estimation
> process, i.e. in a CoreMSMEstimator. Then we can leave all non-core
> timeframes with -1 and keep the time assignment. The CoreMSMEstimator
> needs to change a few things e.g. in the frame sampling to exclude the
> -1 states.

I don't see we would need a new estimator. The time information is not lost, if the out of core states are not in the connected set.
If we would go for a new estimator, which algorithm should it then use? We would have to reproduce/wrap all the other estimators just for that feature,
which grows the API a lot.

franknoe commented:
Am 24/01/17 um 13:54 schrieb Martin K. Scherer:
> On 01/24/2017 01:46 PM, Frank Noe wrote:
> > An alternative is to do the milestoning assignment during the estimation
> > process, i.e. in a CoreMSMEstimator. Then we can leave all non-core
> > timeframes with -1 and keep the time assignment. The CoreMSMEstimator
> > needs to change a few things e.g. in the frame sampling to exclude the
> > -1 states.
>
> I don't see we would need a new estimator. The time information is not 
> lost, if the out of core states are not in the connected set.
> If we would go for a new estimator, which algorithm should it then 
> use? We would have to reproduce/wrap all the other estimators just for 
> that feature,
> which grows the API a lot.
I don't understand this response, but there are arguments for handling 
things on the estimator side, because:

- introducing a fake state really seems wrong, and if you do that in the 
clustering, you have no control what the estimator does with that 
information. Right now you are assuming a certain treatment of 
connectivity, but that is estimator-specific information which shouldn't 
be used at the stage of clustering and is unreliable.
- milestoning count is logically part of the estimation step.

I'm not sure if we need a different estimator or can integrate the 
milestoning/-1 handling in the regular MSM estimator. If we need a new 
estimator, we could just derive from an existing one and would only need 
minor changes.

>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1000#issuecomment-274795519>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMeQoy4vsTD-yQ5gh_VZfaDPNJhlH5uks5rVfR4gaJpZM4LAcej>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

marscher commented:
Agreed about the "fake" state. I've added a "milestone_counting" decorator to wrap the _estimate method to prepare the dtrajs and the offset arrays.

franknoe commented:
Why decorator? Shouldn't it be just an estimation parameter?

We could also decide that if the MSM input dtrajs contain -1, these are 
by default treated with the milestoning algorithm. In fact if there is a 
full partition (no -1 occurs), then the milestoning count is identical 
to the normal count.

Am 24/01/17 um 15:40 schrieb Martin K. Scherer:
>
> Agreed about the "fake" state. I've added a "milestone_counting" 
> decorator to wrap the _estimate method to prepare the dtrajs and the 
> offset arrays.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1000#issuecomment-274821144>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMeQuTWqAb6jJeVuJBnSUjqqlKFl2Wwks5rVg1GgaJpZM4LAcej>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

marscher commented:
It is an estimation parameter:
* default estimation parameter "core_set=None", which means detect the cores automatically, eg. remove -1 states from the beginning and perform milestone_counting within the trajectories. The user can constrain the core_set manually by given a list/array of states.

How should we deal with -1 for HMSM, since this state is used in a different context there?

franknoe commented:
Am 24/01/17 um 15:48 schrieb Martin K. Scherer:
>
> It is an estimation parameter:
>
>   * default estimation parameter "core_set=None", which means detect
>     the cores automatically, eg. remove -1 states from the beginning
>     and perform milestone_counting within the trajectories. The user
>     can constrain the core_set manually by given a list/array of states.
>
OK, then I misunderstood the last message where you said "decorator".
>
> How should we deal with -1 for HMSM, since this state is used in a 
> different context there?
>
I wouldn't mix HMSMs and core MSMs - they are complementary. So we need 
the core_set parameter only for the MaximumLikelihoodMSM and BayesianMSM 
estimators.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1000#issuecomment-274824090>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMeQkeKHUg9GQcPie4m6b3-_8ZKcKVXks5rVg9ZgaJpZM4LAcej>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

marscher commented:
Ok, then we're ready for review/testing. I still need to add some docstrings.

franknoe commented:
OK, I'm happy to review it when Giovanni is fine with it and tests work. 
Maybe @gph82 and @solsson could also give it a try because they were 
playing around with milestoning.

Am 24/01/17 um 15:53 schrieb Martin K. Scherer:
>
> Ok, then we're ready for review/testing. I still need to add some 
> docstrings.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/1000#issuecomment-274825626>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMeQvh9I5xKkn6bPaIcMXC0xwq1PE-rks5rVhBbgaJpZM4LAcej>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

marscher commented:
@gph82, @solsson and @giopina did you have the chance to take a look yet?

hjuinj commented:
hi @marscher , may I ask what is the current status with estimating MSM with milestone counting? I can see that I am able to use the `to_coreset` method in `DiscreteTrajectoryStats` to mark certain states as transition state in my dtrajs, but will I be able to do  `msm.its` and `msm.estimate_markov_model` with the milestoned dtrajs soon? Thank you.

marscher commented:
ITS/MSM estimation should also be possible, because the dtraj is just rewritten to a coring traj prior estimation. I should finish this branch someday.

hjuinj commented:
thank you for the reply. At the moment I get something like this when I do `msm.its()`:
```
RuntimeError: Estimation has failed at ALL lagtimes. Details:
['Error at lag time 1: negative row index found',
 'Error at lag time 2: negative row index found',
 'Error at lag time 5: negative row index found',
 'Error at lag time 10: negative row index found',
 'Error at lag time 20: negative row index found',
 'Error at lag time 50: negative row index found',
 'Error at lag time 100: negative row index found',
 'Error at lag time 200: negative row index found',
 'Error at lag time 300: negative row index found']

```
I think this is because some of the dtrajs can contain -1 values if a particular trajectory starts in the transition state:

marscher commented:
Okay, so the workaround would currently be to first estimate a MSM with the coring and then pass the msm.dtrajs_full to its, because these dtrajs should already been rewritten.

hjuinj commented:
I am sorry, could you please elaborate a bit more? I am not quite sure what you mean by "first estimate a MSM with the coring" ?

marscher commented:
you estimate a simple MaximumlikelihoodMSM and pass your dtrajs, which contain the -1 states. The MSM will rewrite the input to perform the milestone counting and attach the rewritten dtrajs as the attribute "dtrajs_full". You can then reuse this attribute to pass it to ITS.

hjuinj commented:
Thanks again, I tried your suggestion but it seems to run into the same issue. Is this what you meant:
so say initially I have in my `dtrajs` a list of np.arrays like this:
```
[array([0, 0, 0, ..., 5, 5, 5]),
 array([1, 1, 1, ..., 1, 1, 1]),
 array([3, 3, 3, ..., 1, 1, 1]),
 array([1, 1, 1, ..., 1, 1, 1])]
```
And 0 labels transition states, so I do coring via `DiscreteTrajectoryStats(dtrajs).to_coreset([1,2,3,4,5])` ( say 5 is the max label in this case) and my `dtrajs` now looks like this
```
[array([-1, -1, -1, ...,  5,  5,  5]),
 array([1, 1, 1, ..., 1, 1, 1]),
 array([3, 3, 3, ..., 1, 1, 1]),
 array([1, 1, 1, ..., 1, 1, 1])]
```

which has the -1 states. Now if i do:
```python
model = msm.MaximumLikelihoodMSM()
model.estimate(dtrajs)
```
I still get `ValueError: negative row index found`

marscher commented:
May I ask which version of PyEMMA you are using?

hjuinj commented:
Ah I am on 2.5 when I do `pyemma.__version__`, should I have upgraded to 2.5.1?

marscher commented:
Updates are always good to install, but right now for this particular problem it won't help. I thought you are using this branch here. So I was slightly confused about the outcome. I think you would then have to wait until this pull request is finally merged.

hjuinj commented:
Sorry for the confusion, I can try to install your `marscher:milestoning` branch. I just wasn't sure whether it was ready. So using the same steps as I listed above should be enough with your branch?

marscher committed:
[msm] added core set counting

marscher commented:
updated the branch to reflect recent developments. I would be glad, if somebody could find time to review this, so we can finally integrate it into devel.

marscher committed:
missing file

marscher commented:
@giopina do you mind having a look as well, please?

marscher committed:
fix

marscher committed:
fix warning about invalid escape sequences

marscher committed:
fix docstring

marscher committed:
removed obsolete tests (moved to test_discrete_trajectories)

marscher committed:
fix parameter passing.

marscher committed:
fix test

marscher committed:
typo

marscher committed:
[milestone-counting] fix empty set condition

marscher committed:
fix test

thempel commented on pyemma/msm/estimators/_milestone_counting_decorator.py:
```diff
@@ -0,0 +1,71 @@
+        dtraj_offsets  = getattr(self, _OFFSETS_ATTR_NAME)
+        if isinstance(indices, np.ndarray) and indices.dtype == np.int_:
+            _apply_offsets_to_samples(indices, dtraj_offsets)
+        elif isinstance(indices, list) or (isinstance(indices, np.ndarray) and indices.dtype == np.object_):
```
I suppose this doesn't happen because it's a function not public, but what about the case 
```python
isinstance(indices, list) or (isinstance(indices, np.ndarray) and indices.dtype == np.int_ and len(indices.shape) > 1)
```
i.e. an array of arrays instead of a list of arrays? Maybe just raise if conditions are not met?

thempel commented on pyemma/msm/estimators/bayesian_msm.py:
```diff
@@ -132,6 +132,10 @@ def __init__(self, lag=1, nsamples=100, nsteps=None, reversible=True,
             evaluates to 1/nstates.
 
+	 core_set : None (default) or array like, dtype=int
+		    * If set to None, replaces the state -1 and performs a milestone counting. If no state is
```
sentence? "If set to None, replaces state -1 (if applicable) and performs milestone counting. No effect for Voronoi-discretized trajectories (default)."

thempel commented on pyemma/util/discrete_trajectories.py:
```diff
@@ -222,6 +224,163 @@ def number_of_states(dtrajs, only_used = False):
+    core_set: None or ndarray(dtype=int) or list of ndarrays(dtype=int)
+        * If None is given, all indices except -1 are considered to be core sets.
+        * Pass an array of micro-states to define the core sets.
+
```
From this, I don't understand what happends if "list of ndarrays(dtype=int)" is passed. Also not sure if this case is handled as I would guess...

marscher committed:
fix import

thempel commented on pyemma/util/tests/test_discrete_trajectories.py:
```diff
@@ -199,5 +196,116 @@ def test_sample_by_state_replace_subset(self):
+        n_states = 10
+        n_traj = 10
+        dtrajs = [np.random.randint(0, n_states, size=1000) for _ in range(n_traj)]
+        core_set = np.random.randint(0, n_states, size=30)
```
30 states out of 10 states will be assigned core sets? does that make sense?

thempel replied:
btw, how are redundant states handled, i.e. if user passes `core_set = [0, 0, 1]`?

thempel reviewed (commented):
Great, thanks for taking care of this (again)! I'm probably not qualified to review the algorithmic part, but here are some comments.

thempel commented on pyemma/msm/tests/test_msm.py:
```diff
@@ -1094,5 +1095,43 @@ def test_timescales(self):
+
+        msm.pcca(2)
+        samples = msm.sample_by_distributions(msm.metastable_distributions, 3)
+
```
Any checks on `samples`?

thempel committed:
[msm] add milestone counting tests, comparison to hmm, doublewell

marscher committed:
fix warning

marscher committed:
fix docstring

marscher committed:
review

marscher committed:
[msm] moved from decorator pattern to more explicit impl

In OOM-MSM the estimation only uses dtrajs[0:-tau] frames,
we need to take care of it somehow...

marscher committed:
[oom-msm-tests] fix dtraj related tests

thempel committed:
removed non-sense

marscher committed:
merged devel

thempel committed:
[msm] milestone counting; remove assignment from dtraj checking

thempel committed:
[msm] milestone counting, added state assignment last_core method

marscher committed:
[msm] milstone counting, fix unit tests, docs, minor

thempel committed:
extend core msm testing

thempel committed:
unify unassigned frame assignment

thempel committed:
check core set msm estimator compatibility

thempel committed:
add core set msm to stat dist constraint compatibility

thempel committed:
add core set msm unit tests

thempel committed:
added core set MSMs to convenience API

thempel committed:
add tests for core set AMMs

thempel committed:
bugfixes

thempel committed:
add dtraj check for elem < -1, fix test

thempel committed:
Revert "[oom-msm-tests] fix dtraj related tests"

This reverts commit ea824c14060898920485fc705f642f1a667af2a2.
This commit apparently broke unit tests in devel version of
OOM estimator

thempel committed:
[cmsm] add docstrings to convenience functions

thempel committed:
[cmsm-test] check if raises for not implemented

thempel committed:
Merge branch 'devel' into milestoning

thempel committed:
[cmsm] cleanup, fix tests, update authors

marscher commented:
We should add a note to the changelog and then we are good to go IMO. What do you think @thempel?

marscher committed:
[doc] amend changelog

marscher committed:
[conda-recipe] try to unify pip invocation

marscher commented:
@thempel shall we bind the core_set parameter to the implied_timescales api as well?

thempel commented:
> shall we bind the core_set parameter to the implied_timescales api as well?

Isn't that already implemented? 
```python
def timescales_msm(dtrajs, lags=None, nits=None, reversible=True, connected=True, weights='empirical',
                   errors=None, nsamples=50, n_jobs=None, show_progress=True, mincount_connectivity='1/n',
                   only_timescales=False, core_set=None, milestoning_method='last_core'):
   ...
```
